retrofit/src/main/java/retrofit/NetworkCall.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (C) 2015 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package retrofit;
+
+import com.squareup.okhttp.OkHttpClient;
+import com.squareup.okhttp.Request;
+import com.squareup.okhttp.ResponseBody;
+import java.io.IOException;
+import retrofit.converter.Converter;
+
+final class NetworkCall<T> implements Call<T> {
+  private final Endpoint endpoint;
+  private final Converter converter;
+  private final OkHttpClient client;
+  private final MethodInfo methodInfo;
+  private final Object[] args;
+
+  private volatile com.squareup.okhttp.Call rawCall;
+  private boolean executed; // Guarded by this.
+
+  NetworkCall(Endpoint endpoint, Converter converter, OkHttpClient client, MethodInfo methodInfo,
+      Object[] args) {
+    this.endpoint = endpoint;
+    this.converter = converter;
+    this.client = client;
+    this.methodInfo = methodInfo;
+    this.args = args;
+  }
+
+  @SuppressWarnings("CloneDoesntCallSuperClone") // We are a final type & this saves clearing state.
+  @Override public NetworkCall<T> clone() {
+    return new NetworkCall<>(endpoint, converter, client, methodInfo, args);
+  }
+
+  public void enqueue(final Callback<T> callback) {
+    synchronized (this) {
+      if (executed) throw new IllegalStateException("Already executed");
+      executed = true;
+    }
+
+    com.squareup.okhttp.Call rawCall;
+    try {
+      rawCall = createRawCall();
+    } catch (Throwable t) {
+      callback.failure(t);
+      return;
+    }
+    this.rawCall = rawCall;
+
+    rawCall.enqueue(new com.squareup.okhttp.Callback() {
+      private void callFailure(Throwable e) {
+        try {
+          callback.failure(e);
+        } catch (Throwable t) {
+          t.printStackTrace();
+        }
+      }
+
+      private void callSuccess(Response<T> response) {
+        try {
+          callback.success(response);
+        } catch (Throwable t) {
+          t.printStackTrace();
+        }
+      }
+
+      @Override public void onFailure(Request request, IOException e) {
+        callFailure(e);
+      }
+
+      @Override public void onResponse(com.squareup.okhttp.Response rawResponse) {
+        final Response<T> response;
+        try {
+          response = parseResponse(rawResponse);
+        } catch (final Throwable e) {
+          callFailure(e);
+          return;
+        }
+        callSuccess(response);
```

Comment:
I would have expected no need to trampoline off of this thread in the executor delegate.